### PR TITLE
Jwt support

### PR DESCRIPTION
**Pull request by Olena-Pro**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -19,6 +19,10 @@
 		<lombok.mapstruct.binding.version>0.2.0</lombok.mapstruct.binding.version>
 		<mapstruct.version>1.5.5.Final</mapstruct.version>
 
+
```
Comment (Elena-Bruyako) [inline pom.xml]: remove blank lines

File: pom.xml [hunk 2]
```diff
@@ -105,6 +109,26 @@
...
+			<version>${jjwt.version}</version>
+			<scope>runtime</scope>
+		</dependency>
+		<dependency>
+			<groupId>io.jsonwebtoken</groupId>
+			<artifactId>jjwt-jackson</artifactId>
+			<version>${jjwt.version}</version>
+			<scope>runtime</scope>
+		</dependency>
+
```
Comment (Elena-Bruyako) [inline pom.xml]: remove blank lines

File: src/main/java/mate/academy/bookstore/config/OpenApiConfig.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import io.swagger.v3.oas.models.Components;
+import io.swagger.v3.oas.models.OpenAPI;
+import io.swagger.v3.oas.models.security.SecurityRequirement;
+import io.swagger.v3.oas.models.security.SecurityScheme;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class OpenApiConfig {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/config/OpenApiConfig.java:13]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/config/OpenApiConfig.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    @Bean
+    public OpenAPI customOpenApi() {
+        return new OpenAPI()
+                .components(new Components().addSecuritySchemes("BearerAuth",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/config/OpenApiConfig.java]: make BearerAuth constant

File: src/main/java/mate/academy/bookstore/controller/AuthenticationController.java [hunk 5]
```diff
@@ -18,11 +21,17 @@
...
 
     @PostMapping("/registration")
     @ResponseStatus(HttpStatus.OK)
     public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto requestDto)
             throws RegistrationException {
         return userService.register(requestDto);
     }
+
+    @PostMapping("/login")
+    public UserLoginResponseDto login(@RequestBody UserLoginRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/AuthenticationController.java]:
```suggestion
    public UserLoginResponseDto login(@RequestBody @Valid UserLoginRequestDto requestDto) {
```

File: src/main/java/mate/academy/bookstore/controller/AuthenticationController.java [hunk 6]
```diff
@@ -18,11 +21,17 @@
...
+    private final AuthenticationService authenticationService;
 
     @PostMapping("/registration")
     @ResponseStatus(HttpStatus.OK)
     public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto requestDto)
             throws RegistrationException {
         return userService.register(requestDto);
     }
+
+    @PostMapping("/login")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/AuthenticationController.java:35]: Add `@Operation`

File: src/main/java/mate/academy/bookstore/controller/AuthenticationController.java [hunk 7]
```diff
@@ -18,11 +21,17 @@
 @RequestMapping("/auth")
 public class AuthenticationController {
     private final UserService userService;
+    private final AuthenticationService authenticationService;
 
     @PostMapping("/registration")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/controller/AuthenticationController.java:27]: Add `@Operation`

File: src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package mate.academy.bookstore.dto.user;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotEmpty;
+import org.hibernate.validator.constraints.Length;
+
+public record UserLoginRequestDto(
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+        @Length(min = 8,max = 35)
+        @Email
+        String email,
+        @NotEmpty
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java]:
```suggestion
        @NotBlank
```

File: src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package mate.academy.bookstore.dto.user;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotEmpty;
+import org.hibernate.validator.constraints.Length;
+
+public record UserLoginRequestDto(
+        @NotEmpty
+        @Length(min = 8,max = 35)
+        @Email
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java:10]: It's better to create custom Email annotation and validator for it. 
Read, what exactly check `@Email`

File: src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+            Authentication authentication = new UsernamePasswordAuthenticationToken(userDetails,
+                    null, userDetails.getAuthorities());
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java]: Use constant for "Bearer "

File: src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+            UserDetails userDetails = userDetailsService.loadUserByUsername(username);
+            Authentication authentication = new UsernamePasswordAuthenticationToken(userDetails,
+                    null, userDetails.getAuthorities());
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java]: Spring already provides constant for Authorization  - HttpHeaders.AUTHORIZATION

File: src/main/java/mate/academy/bookstore/security/JwtUtil.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+                .compact();
+    }
+
+    public boolean isValidToken(String token) {
+        try {
+            Jws<Claims> claimsJws = Jwts.parserBuilder()
+                    .setSigningKey(secret)
+                    .build()
+                    .parseClaimsJws(token);
+            return !claimsJws.getBody().getExpiration().before(new Date());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/JwtUtil.java]:
```suggestion
            return !getClaimFromToken(token, Claims::getExpiration).before(new Date());
```
Don't need duplicate code Jws<Claims> claimsJws = Jwts.parserBuilder()

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+                    null, userDetails.getAuthorities());
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith(START_WITH)) {
+            return bearerToken.substring(7);
```
Comment (koval-kn) [inline src/main/java/mate/academy/bookstore/security/JwtAuthenticationFilter.java]: redundant check, magic number
```suggestion
        if (bearerToken.startsWith(START_WITH)) {
            return bearerToken.substring(START_WITH.length());
```

**Review by koval-kn — CHANGES_REQUESTED** [review]
please fix prev comments too

File: src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java (hunk 10 above)
Comment (Olena-Pro) [reply src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java:10]: I'm sorry I didn't understand what I should do

File: src/main/java/mate/academy/bookstore/security/JwtUtil.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import java.nio.charset.StandardCharsets;
+import java.security.Key;
+import java.util.Date;
+import java.util.function.Function;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JwtUtil {
+
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/security/JwtUtil.java:16]:
```suggestion
```
don't leave emtpy lines after method/class definition

File: src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+import jakarta.validation.constraints.NotBlank;
+import org.hibernate.validator.constraints.Length;
+
+public record UserLoginRequestDto(
+        @NotBlank
+        @Length(min = 8,max = 35)
+        @Email
+        String email,
+        @NotBlank
+        @Length(min = 8,max = 35)
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/dto/user/UserLoginRequestDto.java:13]:
```suggestion
        @Length(min = 8, max = 35)
```
don't forget to press Ctrl Alt L regularly

**Review by okuzan — APPROVED** [review]

**Pull request merged**